Build failure: tp_smapi

### Steps To Reproduce

Steps to reproduce the behavior:
1. run `sudo nixos-rebuild switch --flake git+https://git.sual.in/casual/nixos-devices.git` on old thinkpad

### Build log

```
sudo nixos-rebuild switch --flake git+https://git.sual.in/casual/nixos-devices.git

building the system configuration...
error: builder for '/nix/store/jvvg6n6k326v2l22i6zx0khffbx7qvwd-tp_smapi-0.43-6.4.1.drv' failed with exit code 2;
       last 10 log lines:
       >       |               ^~~~~~~~~~~
       > /build/tp-smapi-0.43/tp_smapi.c: At top level:
       > /build/tp-smapi-0.43/tp_smapi.c:115:8: warning: 'DEFINE_SEMAPHORE' defined but not used [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wunused-variable-Wunused-variable8;;]
       >   115 | static DEFINE_SEMAPHORE(smapi_mutex);
       >       |        ^~~~~~~~~~~~~~~~
       > cc1: some warnings being treated as errors
       > make[2]: *** [/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/scripts/Makefile.build:252: /build/tp-smapi-0.43/tp_smapi.o] Error 1
       > make[1]: *** [/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/Makefile:2024: /build/tp-smapi-0.43] Error 2
       > make[1]: Leaving directory '/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/build'
       > make: *** [Makefile:46: modules] Error 2
       For full logs, run 'nix log /nix/store/jvvg6n6k326v2l22i6zx0khffbx7qvwd-tp_smapi-0.43-6.4.1.drv'.
error: 1 dependencies of derivation '/nix/store/qv0nlq4904ymx1gsz1ywx5iysh0xa6hg-kernel-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/ffl8bnics12ycmy7czml17a7cjxln97g-linux-6.4.1-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/gp6f39px43r6nhxq8y3dy72fcg69yj06-nixos-system-Thinking-23.11.20230706.3c74875.drv' failed to build



[root@Thinking:/home/casual]# nix log /nix/store/jvvg6n6k326v2l22i6zx0khffbx7qvwd-tp_smapi-0.43-6.4.1.drv


@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/kpnl1b6qvi38gy7blh0k68gzdfzh08wq-tp-smapi-0.43
source root is tp-smapi-0.43
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
configuring
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
building
build flags: -j4 SHELL=/nix/store/51sszqz1d9kpx480scb1vllc00kxlx79-bash-5.2-p15/bin/bash KBASE=/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1->
make -C /nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/build M=/build/tp-smapi-0.43 O=/nix/store/l45kxrj5wsc71lz3mp>
make[1]: Entering directory '/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/build'
  CC [M]  /build/tp-smapi-0.43/thinkpad_ec.o
  CC [M]  /build/tp-smapi-0.43/tp_smapi.o
  CC [M]  /build/tp-smapi-0.43/hdaps.o
/build/tp-smapi-0.43/thinkpad_ec.c:94:42: error: macro "DEFINE_SEMAPHORE" requires 2 arguments, but only 1 given
   94 | static DEFINE_SEMAPHORE(thinkpad_ec_mutex);
      |                                          ^
In file included from /build/tp-smapi-0.43/thinkpad_ec.c:45:
/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/include/linux/semaphore.h:34: note: macro "DEFINE_SEMAPHORE" >
   34 | #define DEFINE_SEMAPHORE(_name, _n)     \
      |
/build/tp-smapi-0.43/thinkpad_ec.c:94:8: error: type defaults to 'int' in declaration of 'DEFINE_SEMAPHORE' [-Werror=implicit-int]
   94 | static DEFINE_SEMAPHORE(thinkpad_ec_mutex);
      |        ^~~~~~~~~~~~~~~~
/build/tp-smapi-0.43/thinkpad_ec.c: In function 'thinkpad_ec_lock':
/build/tp-smapi-0.43/thinkpad_ec.c:112:35: error: 'thinkpad_ec_mutex' undeclared (first use in this function); did you mean 'thinkpad_ec_lock'?
  112 |         ret = down_interruptible(&thinkpad_ec_mutex);
      |                                   ^~~~~~~~~~~~~~~~~
      |                                   thinkpad_ec_lock
/build/tp-smapi-0.43/thinkpad_ec.c:112:35: note: each undeclared identifier is reported only once for each function it appears in
/build/tp-smapi-0.43/thinkpad_ec.c: In function 'thinkpad_ec_try_lock':
/build/tp-smapi-0.43/thinkpad_ec.c:126:30: error: 'thinkpad_ec_mutex' undeclared (first use in this function); did you mean 'thinkpad_ec_lock'?
  126 |         return down_trylock(&thinkpad_ec_mutex);
      |                              ^~~~~~~~~~~~~~~~~
      |                              thinkpad_ec_lock
/build/tp-smapi-0.43/thinkpad_ec.c: In function 'thinkpad_ec_unlock':
/build/tp-smapi-0.43/thinkpad_ec.c:138:13: error: 'thinkpad_ec_mutex' undeclared (first use in this function); did you mean 'thinkpad_ec_lock'?
  138 |         up(&thinkpad_ec_mutex);
      |             ^~~~~~~~~~~~~~~~~
      |             thinkpad_ec_lock
/build/tp-smapi-0.43/thinkpad_ec.c: In function 'thinkpad_ec_try_lock':
/build/tp-smapi-0.43/thinkpad_ec.c:127:1: error: control reaches end of non-void function [-Werror=return-type]
  127 | }
      | ^
/build/tp-smapi-0.43/thinkpad_ec.c: At top level:
/build/tp-smapi-0.43/thinkpad_ec.c:94:8: warning: 'DEFINE_SEMAPHORE' defined but not used [-Wunused-variable]
   94 | static DEFINE_SEMAPHORE(thinkpad_ec_mutex);
      |        ^~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[2]: *** [/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/scripts/Makefile.build:252: /build/tp-smapi-0.4>
make[2]: *** Waiting for unfinished jobs....
/build/tp-smapi-0.43/tp_smapi.c:115:36: error: macro "DEFINE_SEMAPHORE" requires 2 arguments, but only 1 given
  115 | static DEFINE_SEMAPHORE(smapi_mutex);
      |                                    ^
In file included from /nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/include/linux/fs.h:25,
                 from /nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/include/linux/proc_fs.h:10,
                 from /build/tp-smapi-0.43/tp_smapi.c:41:
/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/include/linux/semaphore.h:34: note: macro "DEFINE_SEMAPHORE" >
   34 | #define DEFINE_SEMAPHORE(_name, _n)     \
      |
/build/tp-smapi-0.43/tp_smapi.c:115:8: error: type defaults to 'int' in declaration of 'DEFINE_SEMAPHORE' [-Werror=implicit-int]
  115 | static DEFINE_SEMAPHORE(smapi_mutex);
      |        ^~~~~~~~~~~~~~~~
/build/tp-smapi-0.43/tp_smapi.c: In function 'store_battery_start_charge_thresh':
/build/tp-smapi-0.43/tp_smapi.c:780:15: error: 'smapi_mutex' undeclared (first use in this function)
  780 |         down(&smapi_mutex);
      |               ^~~~~~~~~~~
/build/tp-smapi-0.43/tp_smapi.c:780:15: note: each undeclared identifier is reported only once for each function it appears in
/build/tp-smapi-0.43/tp_smapi.c: In function 'store_battery_stop_charge_thresh':
/build/tp-smapi-0.43/tp_smapi.c:822:15: error: 'smapi_mutex' undeclared (first use in this function)
  822 |         down(&smapi_mutex);
      |               ^~~~~~~~~~~
/build/tp-smapi-0.43/tp_smapi.c: At top level:
/build/tp-smapi-0.43/tp_smapi.c:115:8: warning: 'DEFINE_SEMAPHORE' defined but not used [-Wunused-variable]
  115 | static DEFINE_SEMAPHORE(smapi_mutex);
      |        ^~~~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make[2]: *** [/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/scripts/Makefile.build:252: /build/tp-smapi-0.4>
make[1]: *** [/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/source/Makefile:2024: /build/tp-smapi-0.43] Error 2
make[1]: Leaving directory '/nix/store/l45kxrj5wsc71lz3mpap21mf7bk5ihrs-linux-6.4.1-dev/lib/modules/6.4.1-zen1/build'
make: *** [Makefile:46: modules] Error 2
```

### Additional context

Device: Thinkpad X220

### Notify maintainers

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.3.1-zen1, NixOS, 23.11 (Tapir), 23.11.20230523.d30264c`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

```

--- a/pkgs/os-specific/linux/tp_smapi/default.nix
+++ b/pkgs/os-specific/linux/tp_smapi/default.nix
@@ -1,4 +1,15 @@
-{ stdenv, lib, fetchFromGitHub, kernel, writeScript, coreutils, gnugrep, jq, curl, common-updater-scripts, runtimeShell
+{ stdenv
+, lib
+, fetchFromGitHub
+, fetchpatch
+, kernel
+, writeScript
+, coreutils
+, gnugrep
+, jq
+, curl
+, common-updater-scripts
+, runtimeShell
 }:
 
 stdenv.mkDerivation rec {
@@ -6,12 +17,20 @@ stdenv.mkDerivation rec {
   version = "0.43";
 
   src = fetchFromGitHub {
-    owner = "evgeni";
+    owner = "linux-thinkpad";
     repo = "tp_smapi";
     rev = "tp-smapi/${version}";
     sha256 = "1rjb0njckczc2mj05cagvj0lkyvmyk6bw7wkiinv81lw8m90g77g";
-    name = "tp-smapi-${version}";
   };
+
+  patches = [
+    # update DEFINE_SEMAPHORE usage for linux 6.4+
+    # https://github.com/linux-thinkpad/tp_smapi/pull/45
+    (fetchpatch {
+      url = "https://github.com/linux-thinkpad/tp_smapi/commit/0c3398b1acf2a2cabd9cee91dc3fe3d35805fa8b.patch";
+      hash = "sha256-J/WvijrpHGwFOZMMxnHdNin5eh8vViTcNb4nwsCqsLs=";
+    })
+  ];
 
   nativeBuildInputs = kernel.moduleBuildDependencies;
 
@@ -39,7 +58,7 @@ stdenv.mkDerivation rec {
 
   meta = {
     description = "IBM ThinkPad hardware functions driver";
-    homepage = "https://github.com/evgeni/tp_smapi";
+    homepage = "https://github.com/linux-thinkpad/tp_smapi";
     license = lib.licenses.gpl2;
     maintainers = [ ];
     # driver is only ment for linux thinkpads i think  bellow platforms should cover it.

